docs/flip/2434-general-metadata.md

```diff
@@ -0,0 +1,132 @@
+# FLIP: Axis Metadata
+
+
+- Start Date: 2022-08-08
+- FLIP Issue: [#2434](https://github.com/google/flax/issues/2434)
+- FLIP PR: [#2435](https://github.com/google/flax/pull/2435)
+- Status: Proposal
+
+
+## Summary
+
+This FLIP proposes to extend Flax's variable collections with a generic axis metadata API. 
+The core of the API is a abstract base class that is recognized by lifting transformations that can add an axis (vmap, scan).
+Users can extend the base class to keep track of per-axis metadata in a way that works with lifted transformations.
+
+
+## Motivation
+
+Generally, there is no way in Flax to track metadata for variables across lifted transformations. Axis metadata is used to keep track of semantic information about axes into other (Flax independent) APIs.
+For example, optimizers like AdaFactor can be configured on a per-axis level and partitioning APIs in JAX like xmap or pjit require per variable annotations to map effectiently to parallel hardware.
+
+Currently, there is experimental support for partitioning annotations which requires using dedicated wrapper around lifted transforms that change axes (``nn.scan``, ``nn.vmap``) and a special APIs to create variables (``param_with_axes`` and ``variable_with_axes``). 
+The experimental partitioning API stores the metadata in a seperate collection named "[collection]_axes".
+
+The experimental API has a number of shortcomings that we like to solve:
+1. The current API works for tracking PartitionSpecs but not for other types of metadata like optimizer annotations.
+2. The implementation using an "xxx_axes" collection requires error-prone and non-composable string manipulation.
+3. Special, partioning-aware variable creators and lifted transforms are required
+4. The partioning API is hard to use with pre-existing Modules that aren't partioning aware.
+
+
+## Proposal
+
+To generalize metadata tracking and keep the specific metadata out of core Flax we propose the following abstract base class:
+
+```python
+class AxisMetadata(metaclass=abc.ABCMeta):
+
+  @abc.abstractmethod
+  def unbox(self) -> Any:
+    pass
+
+  @abc.abstractmethod
+  def add_axis(self, index: int, params: Dict[Any, Any]) -> TAxisMetadata:
```

Comment:
What is not entirely clear to me is how `add_axis` and `remove_axis` will be implemented in classes implementing `AxisMetadata`. The way I understand it is that lifted transforms can call these functions to add additional axes where they can store data (for instance of a lifted scan). Are there other use cases for `add_axis`? Maybe you could give an example implementation to these functions just to make things even more clear?

Reply:
Added reference implementation in the Partitioned example below